posts/2023-10-26-broken-badges-and-23k-keywords.md

```diff
@@ -0,0 +1,40 @@
+
+As the quote above already mentions, this crate is using features… a lot of features… almost 23,000! 😱
+
+What crate needs that many features? Well, this crate provides SVG icons for Rust-based web applications… and it uses one feature per icon so that the payload size of the final WebAssembly bundle stays small.
+
+At first glance there should be nothing wrong with this. This seems like a reasonable thing to do from a crate author perspective and neither cargo, nor crates.io, were showing any warnings about this. Unfortunately, some of the internals are not too happy about such a high number of features…
+
+The first problem that was already identified by the crate author: the API responses from crates.io are getting veeeery large. Adding to the problem is the fact that the crates.io API currently does not paginate the list of published versions. Changing this is obviously a breaking change, so our team had been a bit reluctant to change the behavior of the API in that regard, though this situation has shown that we will likely have to tackle this problem in the near future.
+
+The next problem is that the [index file](https://index.crates.io/ic/on/icondata) for this crate is also getting large. With 9 published versions it already contains 11 MB of data. And just like the crates.io API, there is currently no pagination built into the package index file format.
+
+Now you may ask, why do the package index and cargo need to know about features? Well, the easy answer is: for dependency resolution. Features can enable optional dependencies, so when a dependency feature is used it might influence the dependency resolution. Our initial thought was that we could at least drop all empty feature declarations from the index file (e.g. `foo = []`), but the cargo team informed us that cargo relies on them being available there too, and so for backwards-compatibility reasons this is not an option.
+
+On the bright side, most Rust users are on cargo versions these days that use the sparse package index by default, which only downloads index files for packages actually being used. In other words: only users of this icon crate need to pay the price for downloading all the metadata. On the flipside, this means users who are still using the git-based index are all paying for this one crate using 23,000 features.
+
+So, where do we go from here? 🤔
+
+While we believe that supporting such a high number of features is conceptually a valid request, with the current implementation details in crates.io and cargo we cannot support this. After analyzing all of these downstream effects from a single crate having that many features, we realized we need some form of restriction on crates.io to keep the system from falling apart.
+
+Now comes the important part: **on 2023-10-16 the crates.io team deployed a change limiting the number of features a crate can have to 300.**
```

Comment:
It seems like a bit of an anti-pattern to bury "the important part" somewhere in the middle of the blog posts. Maybe this merits a TL;DR at the top of the post? (Not a knock on the rest of the post, I think explaining all the context is very useful!)

Also "limiting the number of features a crate can have" is kind of implicit about how/where this limit is enforced. Given that crates are immutable I assume this is meant to fail uploads of new crates/versions meeting this criterion, maybe make that explicit here?

Reply:
it is important, but I'm not sure it's important enough to warrant a tl;dr section. the relevant publishers that have crates with more than 300 features have already gotten exceptions, and everyone else will be largely unaffected.

> Given that crates are immutable I assume this is meant to fail uploads of new crates/versions meeting this criterion

that is correct.

```suggestion
Now comes the important part: **on 2023-10-16 the crates.io team deployed a change limiting the number of features a crate can have to 300 for any new crates/versions being published.**
```